File: src/main/java/org/swordess/common/vitool/cmd/AliyunOssCommands.java
package org.swordess.common.vitool.cmd;

import com.aliyuncs.DefaultAcsClient;
import com.aliyuncs.exceptions.ClientException;
import com.aliyuncs.http.MethodType;
import com.aliyuncs.profile.DefaultProfile;
import com.aliyuncs.profile.IClientProfile;
import com.aliyuncs.sts.model.v20150401.AssumeRoleRequest;
import com.aliyuncs.sts.model.v20150401.AssumeRoleResponse;
import org.springframework.shell.standard.ShellComponent;
import org.springframework.shell.standard.ShellMethod;

@ShellComponent
public class AliyunOssCommands {

    @ShellMethod("Verify the STS configuration by retrieving the AK.")
    public String aliyunOssVerifySts(String region, String accessKeyId, String accessKeySecret, String arn) {
        try {
            IClientProfile profile = DefaultProfile.getProfile(region, accessKeyId, accessKeySecret);
            DefaultAcsClient client = new DefaultAcsClient(profile);

            AssumeRoleRequest request = new AssumeRoleRequest();
            request.setMethod(MethodType.POST);
            request.setRoleArn(arn);
            request.setRoleSessionName("vitool_verify_" + System.currentTimeMillis());
            request.setDurationSeconds(1000L); // 设置凭证有效时间

            AssumeRoleResponse response = client.getAcsResponse(request);

            String message = "\tExpiration: " + response.getCredentials().getExpiration() + "\n" +
                    "\tAccess Key Id: " + response.getCredentials().getAccessKeyId() + "\n" +
                    "\tAccess Key Secret: " + response.getCredentials().getAccessKeySecret() + "\n" +
                    "\tSecurity Token: " + response.getCredentials().getSecurityToken() + "\n" +
                    "\tRequestId: " + response.getRequestId() + "\n";

            return "SUCCESS\n" + message;

        } catch (ClientException e) {
            String errorMessage = "\tError code: " + e.getErrCode() + "\n" +
                    "\tError message: " + e.getErrMsg() + "\n" +
                    "\tRequestId: " + e.getRequestId();

            return "FAILURE\n" + errorMessage;
        }
    }

}

File: src/main/kotlin/org/swordess/common/vitool/cmd/AliyunOssCommands.kt
/*
 * Copyright (c) 2019-2022 Swordess
 *
 * Distributed under MIT license.
 * See file LICENSE for detail or copy at https://opensource.org/licenses/MIT
 */

package org.swordess.common.vitool.cmd

import com.aliyuncs.DefaultAcsClient
import com.aliyuncs.auth.sts.AssumeRoleRequest
import com.aliyuncs.exceptions.ClientException
import com.aliyuncs.http.MethodType
import com.aliyuncs.profile.DefaultProfile
import com.aliyuncs.profile.IClientProfile
import org.springframework.shell.standard.ShellComponent
import org.springframework.shell.standard.ShellMethod

@ShellComponent
class AliyunOssCommands {

    @ShellMethod(key = ["alioss verify-sts"], value = "Verify the STS configuration by retrieving the AK.")
    fun aliyunOssVerifySts(region: String, accessKeyId: String, accessKeySecret: String, arn: String): String {
        return try {
            val profile: IClientProfile = DefaultProfile.getProfile(region, accessKeyId, accessKeySecret)
            val client = DefaultAcsClient(profile)

            val response = client.getAcsResponse(AssumeRoleRequest().apply {
                sysMethod = MethodType.POST
                roleArn = arn
                roleSessionName = "vitool_verify_" + System.currentTimeMillis()
                durationSeconds = 1000L // 设置凭证有效时间
            })

            with(response) {
                """
                SUCCESS
                    Expiration: ${credentials.expiration}
                    Access Key Id: ${credentials.accessKeyId}
                    Access Key Secret: ${credentials.accessKeySecret}
                    Security Token: ${credentials.securityToken}
                    RequestId: $requestId""".trimIndent()
            }

        } catch (e: ClientException) {
            with(e) {
                """
                FAILURE
                    Error code: $errCode
                    Error message: $errMsg
                    RequestId: $requestId""".trimIndent()
            }
        }
    }

}